### DISCUSSION TITLE
2 Floors so far progress

### DISCUSSION THREAD

**Discussion by vlaryn** [ExperienceLovelace/ha-floorplan#205] [Show and tell]
I have created my first and second floors with lights / climate and media. Still working on basement and other new entities coming. Please let me know what you think and if you have any ideas. 

Thanks
Wayne

https://youtu.be/gGMI6GgyHJI

**Comment by OzGav** [comment]
Nice. Rather than a rolling gif on your TV you could just show a picture based on what app is running. I do something similar to that.

I like the blue frames you are using around the rooms when the lights are on. Looks good.

The lamp colour blending is a nice effect. 

Thanks for sharing!

**Comment by sstarcher** [comment]
@vlaryn do you have your code up somewhere so we could learn from your great work?

**Reply by vlaryn** [reply]
I just posted all the code to my paste bin. 

Here are links to my YAML / CSS / and SVG file. The SVG file was created in inscape and needs to be saved as SVG optimized. Its only really posted to help as a reference. 

YAML
https://pastebin.com/aABrkpbV
CSS
https://pastebin.com/BnCj2y8K
SVG
https://pastebin.com/uBEBNtJS

**Reply by vlaryn** [reply]
I tried posting in reply using code option but it loses its formatting.

**Reply by aparker21** [reply]
Thank you for providing these!!  I have been struggling to get the temperature to show and your code helped out a lot!  I also used it to animate a rectangle around my temperature to show the thermostat state

**Comment by vlaryn** [comment]
`views:
  - title: Floorplan
    path: floorplan
    icon: mdi:floor-plan
    theme: lovelace-blue-blur
    panel: true
    cards:
      - type: 'custom:floorplan-card'
        full_height: true
        style: >
          ha-card {
              max-width: 150vh;
              margin: 0 auto;
              background: none;
          }
        config:
          image: /local/floor_plan/floor_plan.svg
          stylesheet: /local/floor_plan/floor_plan.css
 
          startup_action:
            - service: floorplan.class_set
              service_data:
                element: first_floor
                class: layer-visible
            - service: floorplan.class_set
              service_data:
                elements:
                  - second_floor
                class: layer-hidden
            - service: floorplan.class_set
              service_data:
                element: first_floor.button
                class: button-on
            - service: floorplan.class_set
              service_data:
                elements:
                  - second_floor.button
                class: button-off
 
          rules:
            - element: second_floor.button
              tap_action:
                - service: floorplan.class_set
                  service_data:
                    element: second_floor
                    class: layer-visible
                - service: floorplan.class_set
                  service_data:
                    elements:
                      - first_floor
                    class: layer-hidden
                - service: floorplan.class_set
                  service_data:
                    element: second_floor.button
                    class: button-on
                - service: floorplan.class_set
                  service_data:
                    elements:
                      - first_floor.button
                    class: button-off
              
            - element: first_floor.button
              tap_action:
                - service: floorplan.class_set
                  service_data:
                    element: first_floor
                    class: layer-visible
                - service: floorplan.class_set
                  service_data:
                    elements:
                      - second_floor
                    class: layer-hidden
                - service: floorplan.class_set
                  service_data:
                    element: first_floor.button
                    class: button-on
                - service: floorplan.class_set
                  service_data:
                    elements:
                      - second_floor.button
                    class: button-off
 
            - entity: climate.my_ecobee
              element: tspan4505
              tap_action:
                action: 'more-info'
              state_action:
                action: call-service
                service: floorplan.style_set
                service_data:
                  element: tspan4505
                  style: |
                    >
                    if( entity.state === "heat" )
                        return "fill: rgb(255,165,0);";
                    if( entity.state === "cool" )
                        return "fill: rgb(25,140,255);";
                    if( entity.state === "off" )
                        return "fill: rgb(255,255,255);";
                    if( entity.state === "heat_cool" )
                        return "fill: rgb(43,255,0);";
            - entity: sensor.my_ecobee_temperature
              hold_action:
                action: 'more-info'
              state_action:
                - service: floorplan.text_set
                  service_data:
                    element: sensor.my_ecobee_temperature
                    text: |
                      >
                      return (entity.state !== undefined) ? entity.state : "unknown";
 
            - entity: climate.den_ac_unit
              element: tspan1214
              tap_action:
                action: 'more-info'
              state_action:
                action: call-service
                service: floorplan.style_set
                service_data:
                  element: tspan1214
                  style: |
                    >
                    if( entity.state === "heat" )
                        return "fill: rgb(255,165,0);";
                    if( entity.state === "cool" )
                        return "fill: rgb(25,140,255);";
                    if( entity.state === "off" )
                        return "fill: rgb(255,255,255);";
                    if( entity.state === "fan_only" )
                        return "fill: rgb(43,255,0);";
            - entity: sensor.den_ir_blaster_temperature
              hold_action:
                action: 'more-info'
              state_action:
                - service: floorplan.text_set
                  service_data:
                    element: sensor.den_ir_blaster_temperature
                    text: |
                      >
                      return (entity.state !== undefined) ? entity.state : "unknown";
 
            - entities:
                - light.den
                - light.foyer_garage
                - light.den_lamp
                - light.den_desk_mood_lights
                - light.foyer_dinning_rm
                - light.kitchen_lights
                - light.living_room_shelly25
                - light.twinkly_tree
                - light.garage
                - light.kitchen_island
                - light.dining_room
                - light.master_bath_main_lights
                - switch.master_bedroom_lights
                - switch.nathans_room
                - light.ethans_room_lights
              tap_action: 'false'
              state_action:
                action: call-service
                service: floorplan.class_set
                service_data: 'light-${entity.state}'
 
            - entities:
                - light.den
                - light.foyer_garage
                - light.den_lamp
                - light.den_desk_mood_lights
                - light.foyer_dinning_rm
                - light.kitchen_lights
                - light.living_room_shelly25
                - light.garage
                - light.kitchen_island
                - light.dining_room
                - light.master_bath_main_lights
                - light.ethans_room_lights
              tap_action: 
                action: 'toggle'
              state_action:
                action: 'call-service'
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('light.', 'light_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "on" )
                        return "display: none;";
                    let hue = 0;
                    let sat = 0;
                    if( entity.attributes.hs_color )
                    {
                        hue = entity.attributes.hs_color[0];
                        sat = entity.attributes.hs_color[1];
                    }
                    if( sat < 10 )
                    {
                      return `
                        display: block;
                        filter:
                          brightness(calc( ${entity.attributes.brightness} / 255));`
                    }
                      return `
                        display: block;
                        filter:
                          sepia(100%)
                          hue-rotate(calc( ${hue}deg - 55deg ))
                          saturate(calc( ${sat}% * 2 ))
                          brightness(calc( ${entity.attributes.brightness} / 255));
                          `
 
            - entities:
                - media_player.den
                - media_player.living_room
              tap_action:
                action: 'more-info'
              state_action:
                action: call-service
                service: floorplan.class_set
                service_data: 'media_player-${entity.state}'
 
            - entities:
                - media_player.den
                - media_player.living_room
              tap_action:
                action: 'more-info'
              hold_action:
                action: navigate
                navigation_path: /lovelace/now-playing
              state_action:
                action: call-service
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('media_player.', 'media_player_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "playing" )
                        return "display: none;";
                      return `
                        display: block;
                          `
 
            - entities:
                - switch.den_shelf
                - switch.master_bedroom_lights
                - switch.nathans_room
              tap_action: 
                action: 'toggle'
              state_action:
                action: 'call-service'
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('switch.', 'switch_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "on" )
                        return "display: none;";
                      return `
                        display: block;
                          `
 
            - entities:
                - cover.garage_door_opener
              tap_action: 
                action: 'toggle'
              state_action:
                action: 'call-service'
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('cover.', 'cover_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "closed" )
                        return "display: none;";
                      return `
                        display: block;
                          `
 
            - entities:
                - person.vlaryn
                - person.rachel
              tap_action: 
                action: 'toggle'
              state_action:
                action: 'call-service'
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('person.', 'person_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "home" )
                        return "display: none;";
                      return `
                        display: block;
                          `
 
            - entities:
                - input_boolean.foyer_toggle
                - input_boolean.masterbathventlight
                - input_boolean.2nd_hallway
                - input_boolean.2nd_landing
                - input_boolean.laundry
                - input_boolean.kids_bath_main
                - input_boolean.kids_bath_mirror
              tap_action: 'false'
              state_action:
                action: call-service
                service: floorplan.class_set
                service_data: 'inputboolean-${entity.state}'
 
            - entities:
                - input_boolean.foyer_toggle
                - input_boolean.masterbathventlight
                - input_boolean.2nd_hallway
                - input_boolean.2nd_landing
                - input_boolean.laundry
                - input_boolean.kids_bath_main
                - input_boolean.kids_bath_mirror
              tap_action: 
                action: 'toggle'
              state_action:
                action: 'call-service'
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('input_boolean.', 'input_boolean_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "on" )
                        return "display: none;";
                    let hue = 0;
                    let sat = 0;
                    if( entity.attributes.hs_color )
                    {
                        hue = entity.attributes.hs_color[0];
                        sat = entity.attributes.hs_color[1];
                    }
                    if( sat < 10 )
                    {
                      return `
                        display: block;
                        filter:
                          brightness(calc( ${entity.attributes.brightness} / 255));`
                    }
                      return `
                        display: block;
                        filter:
                          sepia(100%)
                          hue-rotate(calc( ${hue}deg - 55deg ))
                          saturate(calc( ${sat}% * 2 ))
                          brightness(calc( ${entity.attributes.brightness} / 255));
                          `
 
            - entities:
                - light.twinkly_tree
              tap_action: 
                action: 'toggle'
              state_action:
                action: 'call-service'
                service: floorplan.style_set
                service_data:
                  element: "${entity.entity_id.replace('light.', 'light_special_overlay.')}"
                  style: |
                    >
                    if( entity.state !== "on" )
                        return "display: none;";
                    let hue = 0;
                    let sat = 0;
                    if( entity.attributes.hs_color )
                    {
                        hue = entity.attributes.hs_color[0];
                        sat = entity.attributes.hs_color[1];
                    }
                    if( sat < 10 )
                    {
                      return `
                        display: block;
                        filter:
                          brightness(calc( ${entity.attributes.brightness} / 255));`
                    }
                      return `
                        display: block;
                        filter:
                          sepia(100%)
                          hue-rotate(calc( ${hue}deg - 55deg ))
                          saturate(calc( ${sat}% * 2 ))
                          brightness(calc( ${entity.attributes.brightness} / 255));
                          `
 `

**Reply by OzGav** [reply]
@vlaryn Just looking at this and I notice you have what seems to be unnecessary duplication in there. `light.den` for example you have in two places with the state action of one setting the class and the other setting the style. If that was necessary I would normally have those two service calls together under the one state action. It also means you don't have the odd situation of one section having tap action none and the other toggle. This is what I would do:
```
            - entities:
                - light.twinkly_tree
                - switch.master_bedroom_lights
                - switch.nathans_room
              tap_action: false
              state_action:
                service: floorplan.class_set
                service_data: 'light-${entity.state}'
 
            - entities:
                - light.den
                - light.foyer_garage
                - light.den_lamp
                - light.den_desk_mood_lights
                - light.foyer_dinning_rm
                - light.kitchen_lights
                - light.living_room_shelly25
                - light.garage
                - light.kitchen_island
                - light.dining_room
                - light.master_bath_main_lights
                - light.ethans_room_lights
              tap_action: toggle
              state_action:
                - service: floorplan.class_set
                  service_data: 'light-${entity.state}'
                - service: floorplan.style_set
                  service_data:
...
```

**Reply by vlaryn** [reply]
@OzGav Funny I just noticed this yesterday when looking over my yaml to help someone else out. I was kind of unsure if I could do both a class and style change that way so I am glad you saw it and let me know. I am planning to rewrite the yaml and make it cleaner with section labels etc. I have found several things over the last year and just quickly fixed. 

Wayne B.

**Comment by vlaryn** [comment]
CSS
@keyframes fade-in {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
 
@keyframes flip-in-hor-top {
  0% {
    -webkit-transform: rotateX(-80deg);
            transform: rotateX(-80deg);
    opacity: 0;
  }
  100% {
    -webkit-transform: rotateX(0);
            transform: rotateX(0);
    opacity: 1;
  }
}
 
@keyframes active-pulse {
    0% {
    filter: unset;
    stroke-opacity: 1;
    stroke-width: 1px;
  }
  15% {
    stroke-opacity: 1;
  }
  100% {
    filter: unset;
    stroke-opacity: 0;
    stroke-width: 15px;
  }
}
 
svg, svg * {
  pointer-events: all !important;
  cursor: auto !important;
}
 
/* Layers */
 
.layer-visible {
  display: block !important;
  animation: fade-in 1.2s cubic-bezier(0.390, 0.575, 0.565, 1.000) both;
}
 
.layer-hidden {
  display: none !important;
}
 
/* Buttons */
 
.button-on rect {
  fill: var(--paper-item-icon-active-color) !important;
}
 
/*
.button-off rect {
  fill: #d32f2f !important;
}
*/
 
.button-on tspan,
.button-off tspan {
  fill: white !important;
}
 
svg path {
  fill: inherit;
  stroke: inherit;
  stroke-width: inherit;
}
 
#light-overlays image {
  display: none;
  mix-blend-mode: lighten;
  animation: 0.75s fade-in linear forwards;
}
#light-overlays image.light-on {
  display: block;
}
 
#light-overlays2 image {
  display: none;
  mix-blend-mode: lighten;
  animation: 0.75s fade-in linear forwards;
}
#light-overlays2 image.light-on {
  display: block;
}
 
#person-overlays image {
  display: none;
  mix-blend-mode: normal;
  animation: 0.75s fade-in linear forwards;
}
#person-overlays image.light-on {
  display: block;
}
 
#covers-overlays image {
  display: none;
  mix-blend-mode: normal;
  animation: flip-in-hor-top 7s cubic-bezier(0.250, 0.460, 0.450, 0.940) both;
}
#covers-overlays image.cover-open {
  display: block;
}
 
#areas path {
  fill: none;
}
 
#areas2 path {
  fill: none;
}
 
#areas .light-on {
  stroke-width: 3px;
  stroke: var(--paper-item-icon-active-color);;
}
 
#areas2 .light-on {
  stroke-width: 3px;
  stroke: var(--paper-item-icon-active-color);;
}
 
#areas .inputboolean-on {
  stroke-width: 3px;
  stroke: red;
}
 
#areas2 .inputboolean-on {
  stroke-width: 3px;
  stroke: red;
}
 
g#entities path {
  stroke: rgba(128, 128, 128, 0.521);
  stroke-width: 2px;
  stroke-opacity: 1;
  fill: none;
}
 
g#entities .media_player-idle,
g#entities .media_player-standby,
g#entities .media_player-paused {
  stroke: rgba(218, 206, 47, 0.432);
  stroke-width: 3px;
  opacity: 1;
}
g#entities .media_player-playing {
  stroke: rgba(13, 13, 175, 0.555);
  animation: 2s active-pulse infinite;
}
 
#tv_overlays image {
  display: none;
  mix-blend-mode: normal;
  animation: 2s fade-in both;
}
#tv_overlays image.media_player-playing {
  display: block;
}

**Comment by ferbulous** [comment]
Hi @vlaryn 
Thanks for posting this,
Was able to get switch entities working with the light_overlays thanks to your examples
I just couldn't get the area lines to glow/visible only when the entity's on
Mind taking a look at my area layers, css?

css
https://pastebin.com/iSzrzgAH

svg
https://pastebin.com/K93NCCJ8

yaml
https://pastebin.com/9MQ8Qehx

**Reply by vlaryn** [reply]
Does it work for you lights but not switches? Or does it not work for anything?

**Reply by ferbulous** [reply]
The light & switch overlays works fine.
Just the area lines stays visible the entire time (instead of showing up/glowing when entity is on)

**Reply by vlaryn** [reply]
So if this is an issue with only your switches and not light entities. I think I might have found your issue. Funny this was an issue in my own CSS and yaml as well. So I found in my yaml rule for my switch entities I was doing a call-service / service: floorplan.class_set / service_data: 'light-${entity.state}' instead of 'switch-${entity.state}'. So when I tried to fix mine. I was no longer able to see the lines show up on my own. So that sent me back to my css to see what I had missed. Yep I was missing a statement for #areas .switch-on; . Once I added this I was back to working. For yours I think you just might need to add the following to your CSS file.

CSS
#areas .switch-on {
stroke-width: 3px;
stroke:#a58ee1;
}

Let me know how it works out.
Wayne B.

**Reply by vlaryn** [reply]
I just read your last comment. Are you referring to the pulsing blue lines I have around my TV's when they are playing?

**Reply by vlaryn** [reply]
Oh wait.  You are saying that the lines are always visible.

comment out lines 23-25 and add the following. 

#areas path {
  fill: none;
}

Try adding that.  Not sure you need the stuff at the bottom of your CSS lines 76-96. maybe comment those out as well unless I missed something.

**Reply by ferbulous** [reply]
Thanks @vlaryn , i tried adding those lines but for some reason it's still not working for me
Lines 76-96 are just for my button icons

https://user-images.githubusercontent.com/24526737/211231302-b703bac7-e30c-4bbf-a6d9-8044ed449ce0.mp4

css
```
@keyframes fade-in {
	from {
		opacity: 0;
	}
	to {
		opacity: 1;
	}
}


svg, svg * {
  pointer-events: all !important;
  cursor: auto !important;
}

svg path {
  fill: inherit;
  stroke: inherit;
  stroke-width: inherit;
}


#areas path {
	fill: none;
	}

#areas .light-on {
	stroke-width: 6px;
	stroke-opacity: 1;
	stroke: #a58ee1;
  }

#areas .switch-on {
	stroke-width: 3px;
	stroke:#a58ee1;
}

#light-overlays image {
  display: none;
  mix-blend-mode: lighten;
  animation: 0.75s fade-in linear forwards;
}

#light-overlays image.light-on {
  display: block;
}


#areas .motion-on
{
 	fill: #ff00002e;
 	stroke: #ff000087;
 	stroke-width: 3px;
 	stroke-opacity: 1;
 	filter: unset;
 	animation: fade-in 1s infinite alternate;
}

/* Spinning fan */

.spinning {
	animation-name: spin;
	animation-duration: 5s;
	animation-iteration-count: infinite;
	animation-timing-function: linear;
	transform-origin: center;
	transform-box: fill-box;
  }
  
  @keyframes spin {
	from {
	  transform: rotate(0deg);
	}
	to {
	  transform: rotate(360deg);
	}
  }

/* Button Icons */

#buttons .light-off *  {
  opacity: 0.5 !important;
  stroke-opacity: 0 !important;
}

#buttons .light-on * {
  stroke: #03A9F4 !important;
  opacity: 1 !important;
  stroke-opacity: 1 !important;
}

#buttons .switch-off *  {
  opacity: 0.5 !important;
  stroke-opacity: 0 !important;
}

#buttons .switch-on * {
  stroke: #03A9F4 !important;
  opacity: 1 !important;
  stroke-opacity: 1 !important;
}

/* Temperature text */

.static-temp, .static-temp tspan {
	fill: #ffffff;
	position: fixed;
	transform: skew(213deg) scale(1.5);
  }
```

**Reply by vlaryn** [reply]
![Screenshot 2023-01-08 220551](https://user-images.githubusercontent.com/6608577/211235332-b5eebff0-9199-43a8-9b32-c2c1eb9ae561.png)


Try adding xlink:href    #path200  to this svg:g id="switch.master_bedroom_3g_g1"~~~~~~~  as shown in the picture. I think because you are using buttons to switch on the lights they need to reference the  svg path.

Wayne B.

**Reply by ferbulous** [reply]
@vlaryn oh for that one, i was using switch-x and created another entity for the same device to use with button icons.
Nevermind that, I've removed the button icons and highlighted the area around the tube lights & my night lamp, but I'm scratching my head why it still doesn't work.

Decided to make a new svg file with 2 layers of light areas, but I must be missing something here if I still can't get the area lines to glow.
Do you mind checking this file?

[floorplan_lines.zip](https://github.com/ExperienceLovelace/ha-floorplan/files/10372031/floorplan_lines.zip)

**Reply by OzGav** [reply]
Should you have cropped out the dark areas of the overlaying images?

**Reply by vlaryn** [reply]
@ferbulous So I just starting to look over your data and will post anything I find as I go. I started looking first at your svg file. I am thinking the issue might be there. 1st thing I noticed is in your areas section for both inkscape:label="area.guestroom" and "area.guest_room" your labels look wrong. You need to set the id to <svg:g id="areas.guest_bathroom"> in place of layer5. The inkscape ID is not used by floorplan. Also change layer6 to "areas.guest_room". Be careful to not forget to make areas plural. I did the same thing and had issues when I first made my svg. I am including a zip of 2 files one is a modifed version of your floorplan_lines and the other is an optimized version. I think either will work but I always export mine as an optimized svg.

See if that fixes it.
[corrections.zip](https://github.com/ExperienceLovelace/ha-floorplan/files/10377500/corrections.zip)

**Reply by vlaryn** [reply]
> Should you have cropped out the dark areas of the overlaying images?

I did crop out the dark areas on my overlay images but I am not sure you would need to. The lighten blend mode I don't think will affect it in a negative way.

**Reply by ferbulous** [reply]
> Be careful to not forget to make areas plural. I did the same thing and had issues when I first made my svg

Thanks! @vlaryn  I thought if the elements are under that same areas layer, the id names didn't really matter (area.xx/areas.xx)
Changed it to areas.(room) and I'm getting somewhere 

With developer view enabled
The default area lines are still there by default
The stroke width does change when entity is on, but not the colors (rgb picker is crossed?)

https://user-images.githubusercontent.com/24526737/211462267-d686373a-5f3b-4f47-b300-42e255bf43ef.mp4

**Reply by OzGav** [reply]
Just curious why you guys are using the # CSS selector. Why not just have an element which is the lights on picture and another element on top of that which is a SVG polygon. Then in your rules use the same entity for each of those two room elements so they will both change when the one entity changes state. Then all you need is a state on and a state off CSS class for the picture and the polygon, so just four classes?

**Reply by vlaryn** [reply]
@ferbulous So something is overriding your CSS rgb color. Try adding !important after your rgb value and before the ; for your areas .light-on. see example below.

#areas .light-on {
  stroke-width: 3px;
  stroke: rgb(8, 166, 218) !important;
}

**Reply by vlaryn** [reply]
@OzGav I will look into doing that.  I was very new to this when I started playing around and built mine. I will test some stuff out tomorrow.

**Reply by OzGav** [reply]
Cool. You can link multiple entity and element combinations together in the one rule. See here for an example https://experiencelovelace.github.io/ha-floorplan/docs/example-floorplanner-home/

**Reply by ferbulous** [reply]
@vlaryn hi, so after troubleshooting for a few days, i conclude I have no idea my area layers just won't work.
But I was able to modify the lines on my button icons layers instead, so I'm just gonna stick with this solution.

Also, I noticed one thing, the colors are not correctly displayed on my ios mobile devices (ipad, iphone), works fine on my android device & pc browsers though. Did you have the same issue?

https://user-images.githubusercontent.com/24526737/212626019-06ea9bd7-6757-4eaf-b4f7-cae498025949.MOV

**Reply by OzGav** [reply]
Yes looks like someone else had a problem as well https://github.com/ExperienceLovelace/ha-floorplan/discussions/234

**Reply by ferbulous** [reply]
Thanks, so this could be just an ios bug.

**Reply by OzGav** [reply]
I’m away for a week but I will try and remember to see if I can recreate this issue when I get back.

**Reply by vlaryn** [reply]
@ferbulous So I might have stumbled on a fix for your issue. When you are saving your  SVG file are you doing Save a Copy and changing the type to Optimized? You will want to make sure you do Save a copy as optimized separate from the Saved version you edit. BE careful not overwrite the inkscape format version.  See my quick video for optimized settings.

https://youtu.be/8cxbQ1jwW24
